### PR TITLE
Create a make target to package only the CRD in the recipe bundle

### PR DESCRIPTION
**Pull request by raghavendra-talur**
We need to ship the recipe CRD but without the operator as there is no value is reconciling the recipe CRs for now. Ramen interprets the recipe when kubeObjectProtection is enabled and hands off the resulting backup/restore job to velero.

This PR adds a make target to generate a bundle with no operator in it.

### CODE REVIEW

**Comment by raghavendra-talur** [comment]
Here is the diff between the two bundles that are generated.

```
diff -ruN bundle bundle-crd-only
diff -ruN bundle/manifests/recipe-controller-manager-metrics-service_v1_service.yaml bundle-crd-only/manifests/recipe-controller-manager-metrics-service_v1_service.yaml
--- bundle/manifests/recipe-controller-manager-metrics-service_v1_service.yaml  2024-04-30 16:54:04
+++ bundle-crd-only/manifests/recipe-controller-manager-metrics-service_v1_service.yaml 1969-12-31 19:00:00
@@ -1,23 +0,0 @@
-apiVersion: v1
-kind: Service
-metadata:
-  creationTimestamp: null
-  labels:
-    app.kubernetes.io/component: kube-rbac-proxy
-    app.kubernetes.io/created-by: recipe
-    app.kubernetes.io/instance: controller-manager-metrics-service
-    app.kubernetes.io/managed-by: kustomize
-    app.kubernetes.io/name: service
-    app.kubernetes.io/part-of: recipe
-    control-plane: controller-manager
-  name: recipe-controller-manager-metrics-service
-spec:
-  ports:
-  - name: https
-    port: 8443
-    protocol: TCP
-    targetPort: https
-  selector:
-    control-plane: controller-manager
-status:
-  loadBalancer: {}
diff -ruN bundle/manifests/recipe-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml bundle-crd-only/manifests/recipe-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml
--- bundle/manifests/recipe-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml        2024-04-30 16:54:04
+++ bundle-crd-only/manifests/recipe-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml       1969-12-31 19:00:00
@@ -1,17 +0,0 @@
-apiVersion: rbac.authorization.k8s.io/v1
-kind: ClusterRole
-metadata:
-  creationTimestamp: null
-  labels:
-    app.kubernetes.io/component: kube-rbac-proxy
-    app.kubernetes.io/created-by: recipe
-    app.kubernetes.io/instance: metrics-reader
-    app.kubernetes.io/managed-by: kustomize
-    app.kubernetes.io/name: clusterrole
-    app.kubernetes.io/part-of: recipe
-  name: recipe-metrics-reader
-rules:
-- nonResourceURLs:
-  - /metrics
-  verbs:
-  - get
diff -ruN bundle/manifests/recipe.clusterserviceversion.yaml bundle-crd-only/manifests/recipe.clusterserviceversion.yaml
--- bundle/manifests/recipe.clusterserviceversion.yaml  2024-04-30 16:54:04
+++ bundle-crd-only/manifests/recipe.clusterserviceversion.yaml 2024-04-30 16:53:51
@@ -21,7 +21,7 @@
         }
       ]
     capabilities: Basic Install
-    createdAt: "2024-04-30T20:54:04Z"
+    createdAt: "2024-04-30T20:53:51Z"
     olm.skipRange: ""
     operators.operatorframework.io/builder: operator-sdk-v1.31.0
     operators.operatorframework.io/project_layout: go.kubebuilder.io/v3
@@ -43,200 +43,28 @@
     mediatype: ""
   install:
     spec:
-      clusterPermissions:
-      - rules:
-        - apiGroups:
-          - ramendr.openshift.io
-          resources:
-          - recipes
-          verbs:
-          - create
-          - delete
-          - get
-          - list
-          - patch
-          - update
-          - watch
-        - apiGroups:
-          - ramendr.openshift.io
-          resources:
-          - recipes/finalizers
-          verbs:
-          - update
-        - apiGroups:
-          - ramendr.openshift.io
-          resources:
-          - recipes/status
-          verbs:
-          - get
-          - patch
-          - update
-        - apiGroups:
-          - authentication.k8s.io
-          resources:
-          - tokenreviews
-          verbs:
-          - create
-        - apiGroups:
-          - authorization.k8s.io
-          resources:
-          - subjectaccessreviews
-          verbs:
-          - create
-        serviceAccountName: recipe-controller-manager
-      deployments:
-      - label:
-          app.kubernetes.io/component: manager
-          app.kubernetes.io/created-by: recipe
-          app.kubernetes.io/instance: controller-manager
-          app.kubernetes.io/managed-by: kustomize
-          app.kubernetes.io/name: deployment
-          app.kubernetes.io/part-of: recipe
-          control-plane: controller-manager
-        name: recipe-controller-manager
-        spec:
-          replicas: 1
-          selector:
-            matchLabels:
-              control-plane: controller-manager
-          strategy: {}
-          template:
-            metadata:
-              annotations:
-                kubectl.kubernetes.io/default-container: manager
-              labels:
-                control-plane: controller-manager
-            spec:
-              affinity:
-                nodeAffinity:
-                  requiredDuringSchedulingIgnoredDuringExecution:
-                    nodeSelectorTerms:
-                    - matchExpressions:
-                      - key: kubernetes.io/arch
-                        operator: In
-                        values:
-                        - amd64
-                        - arm64
-                        - ppc64le
-                        - s390x
-                      - key: kubernetes.io/os
-                        operator: In
-                        values:
-                        - linux
-              containers:
-              - args:
-                - --secure-listen-address=0.0.0.0:8443
-                - --upstream=http://127.0.0.1:8080/
-                - --logtostderr=true
-                - --v=0
-                image: gcr.io/kubebuilder/kube-rbac-proxy:v0.13.0
-                name: kube-rbac-proxy
-                ports:
-                - containerPort: 8443
-                  name: https
-                  protocol: TCP
-                resources:
-                  limits:
-                    cpu: 500m
-                    memory: 128Mi
-                  requests:
-                    cpu: 5m
-                    memory: 64Mi
-                securityContext:
-                  allowPrivilegeEscalation: false
-                  capabilities:
-                    drop:
-                    - ALL
-              - args:
-                - --health-probe-bind-address=:8081
-                - --metrics-bind-address=127.0.0.1:8080
-                - --leader-elect
-                command:
-                - /manager
-                image: controller:latest
-                livenessProbe:
-                  httpGet:
-                    path: /healthz
-                    port: 8081
-                  initialDelaySeconds: 15
-                  periodSeconds: 20
-                name: manager
-                readinessProbe:
-                  httpGet:
-                    path: /readyz
-                    port: 8081
-                  initialDelaySeconds: 5
-                  periodSeconds: 10
-                resources:
-                  limits:
-                    cpu: 500m
-                    memory: 128Mi
-                  requests:
-                    cpu: 10m
-                    memory: 64Mi
-                securityContext:
-                  allowPrivilegeEscalation: false
-                  capabilities:
-                    drop:
-                    - ALL
-              securityContext:
-                runAsNonRoot: true
-              serviceAccountName: recipe-controller-manager
-              terminationGracePeriodSeconds: 10
-      permissions:
-      - rules:
-        - apiGroups:
-          - ""
-          resources:
-          - configmaps
-          verbs:
-          - get
-          - list
-          - watch
-          - create
-          - update
-          - patch
-          - delete
-        - apiGroups:
-          - coordination.k8s.io
-          resources:
-          - leases
-          verbs:
-          - get
-          - list
-          - watch
-          - create
-          - update
-          - patch
-          - delete
-        - apiGroups:
-          - ""
-          resources:
-          - events
-          verbs:
-          - create
-          - patch
-        serviceAccountName: recipe-controller-manager
+      deployments: []
     strategy: deployment
   installModes:
-  - supported: false
+  - supported: true
     type: OwnNamespace
-  - supported: false
+  - supported: true
     type: SingleNamespace
   - supported: false
     type: MultiNamespace
-  - supported: true
+  - supported: false
     type: AllNamespaces
   keywords:
   - recipe
   links:
   - name: Recipe
-    url: https://recipe.domain
+    url: https://github.com/ramendr/recipe
   maintainers:
-  - email: your@email.com
-    name: Maintainer Name
+  - email: bmekhiss@redhat.com
+    name: Benamar Mekhissi
+  - email: srangana@redhat.com
+    name: Shyam Ranganathan
   maturity: alpha
   provider:
-    name: Provider Name
-    url: https://your.domain
+    name: Red Hat, Inc.
   version: 0.0.1
```

File: Makefile [hunk 1]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
 	$(KUSTOMIZE) build config/manifests | $(OSDK) generate bundle $(BUNDLE_GEN_FLAGS)
 	$(OSDK) bundle validate ./bundle
 
+.PHONY: bundle-crd-only
+bundle-crd-only: manifests kustomize operator-sdk ## Generate bundle manifests for CRDs only and metadata, then validate generated files.
```
Comment (nirs) [inline Makefile]: Any reason to keep the standard targets with the controller? What is the purpose of the controller anyway?

File: Makefile [hunk 2]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
...
+	$(OSDK) generate kustomize manifests -q --interactive=false --input-dir config/manifests-crd  --output-dir config/manifests-crd
+	cd config/manifests-crd/ && $(KUSTOMIZE) edit add patch --name recipe.v0.0.0 --kind ClusterServiceVersion\
+		--patch '[{"op": "add", "path": "/metadata/annotations/olm.skipRange", "value": "$(SKIP_RANGE)"}]' && \
+		$(KUSTOMIZE) edit add patch --name recipe.v0.0.0 --kind ClusterServiceVersion\
```
Comment (nirs) [inline Makefile]: Adding space after ClusterServiceVersion would be nice

File: Makefile [hunk 3]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
 	$(KUSTOMIZE) build config/manifests | $(OSDK) generate bundle $(BUNDLE_GEN_FLAGS)
 	$(OSDK) bundle validate ./bundle
 
+.PHONY: bundle-crd-only
+bundle-crd-only: manifests kustomize operator-sdk ## Generate bundle manifests for CRDs only and metadata, then validate generated files.
+	$(OSDK) generate kustomize manifests -q --interactive=false --input-dir config/manifests-crd  --output-dir config/manifests-crd
+	cd config/manifests-crd/ && $(KUSTOMIZE) edit add patch --name recipe.v0.0.0 --kind ClusterServiceVersion\
```
Comment (nirs) [inline Makefile]: Adding space after ClusterServiceVersion would be nice.

File: Makefile [hunk 4]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
 	$(KUSTOMIZE) build config/manifests | $(OSDK) generate bundle $(BUNDLE_GEN_FLAGS)
 	$(OSDK) bundle validate ./bundle
 
+.PHONY: bundle-crd-only
+bundle-crd-only: manifests kustomize operator-sdk ## Generate bundle manifests for CRDs only and metadata, then validate generated files.
+	$(OSDK) generate kustomize manifests -q --interactive=false --input-dir config/manifests-crd  --output-dir config/manifests-crd
```
Comment (nirs) [inline Makefile]: It is very hard to work with such long lines of cryptic commands. Lets have *one* argument per line.

File: Makefile [hunk 5]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
...
+	cd config/manifests-crd/ && $(KUSTOMIZE) edit add patch --name recipe.v0.0.0 --kind ClusterServiceVersion\
+		--patch '[{"op": "add", "path": "/metadata/annotations/olm.skipRange", "value": "$(SKIP_RANGE)"}]' && \
```
Comment (nirs) [inline Makefile:246]: Why do we need this patch? it must be documented in above this line.

**Comment by raghavendra-talur** [comment]
Merge after #18

File: Makefile [hunk 6]
```diff
@@ -229,10 +229,25 @@ bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metada
...
+		$(KUSTOMIZE) edit add patch --name recipe.v0.0.0 --kind ClusterServiceVersion\
+		--patch '[{"op": "replace", "path": "/spec/replaces", "value": "$(REPLACES)"}]'
```
Comment (nirs) [inline Makefile:248]: Same for this patch.

File: config/manifests-crd/bases/recipe.clusterserviceversion.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
+apiVersion: operators.coreos.com/v1alpha1
```
Comment (nirs) [inline config/manifests-crd/bases/recipe.clusterserviceversion.yaml:1]: Is this file generated by some tool?

**Review by nirs — COMMENTED** [review]
Can we simplify and replace the targets creating the bundle with the controller with a bundle without the controller? If we don't need the controller, probably nobody else need it.

File: config/manifests-crd/bases/recipe.clusterserviceversion.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  displayName: Recipe
+  icon:
+  - base64data: ""
+    mediatype: ""
+  install:
+    spec:
+      deployments: null
+    strategy: ""
+  installModes:
+  - supported: true
```
Comment (ShyamsundarR) [inline config/manifests-crd/bases/recipe.clusterserviceversion.yaml:33]: I was mildly confused at this scope, but as there is no operator we should be fine.

**Review by ShyamsundarR — APPROVED** [review]
LGTM, I would suggest based on priority of getting fixes in additional comments be handled later as time permits unless critical.

File: config/manifests-crd/bases/recipe.clusterserviceversion.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+spec:
+  apiservicedefinitions: {}
+  customresourcedefinitions:
+    owned:
+    - description: Recipe is the Schema for the recipes API
+      displayName: Recipe
+      kind: Recipe
+      name: recipes.ramendr.openshift.io
+      version: v1alpha1
+  description: Recipe description. TODO.
```
Comment (nirs) [inline config/manifests-crd/bases/recipe.clusterserviceversion.yaml]: We can use: "Recipe describes a workflow used for a capturing or recovering Kubernetes resources."

(based on https://github.com/RamenDR/ramen/blob/main/docs/recipe.md)

**Comment by nirs** [comment]
@raghavendra-talur Tested locally (after rebase on main).

File: Makefile (hunk 1 above)
Comment (raghavendra-talur) [reply Makefile]: We are still working with the IBM team to decide if we want to ship the complete functionality of the recipe controller someday. Will keep the existing target for that future use-case.

As of now, we need a target that bundles only the crd.

File: Makefile (hunk 3 above)
Comment (raghavendra-talur) [reply Makefile]: done

File: Makefile (hunk 2 above)
Comment (raghavendra-talur) [reply Makefile]: Done

File: Makefile (hunk 4 above)
Comment (raghavendra-talur) [reply Makefile]: Done

File: config/manifests-crd/bases/recipe.clusterserviceversion.yaml (hunk 7 above)
Comment (raghavendra-talur) [reply config/manifests-crd/bases/recipe.clusterserviceversion.yaml:1]: This is generated by controller-gen but it should be committed as it has provider info and description that is lost on every generation if not committed.

File: config/manifests-crd/bases/recipe.clusterserviceversion.yaml (hunk 8 above)
Comment (raghavendra-talur) [reply config/manifests-crd/bases/recipe.clusterserviceversion.yaml:33]: This is required as the recipe crd bundle will be installed by the ODF operator in its operator group. The install modes of all the CSVs in the operator group should match. Just copying the installModes of ODF operator here. It has no impact on how the recipe CRD is reconciled as that is decided by the installMode of the recipe controller(CSV).

File: Makefile (hunk 6 above)
Comment (raghavendra-talur) [reply Makefile:248]: done

File: Makefile (hunk 5 above)
Comment (raghavendra-talur) [reply Makefile:246]: done

**Review by nirs — APPROVED** [review]

**Comment by nirs** [comment]
Tested again locally, built bundle looks good to me, and building a bundle container works.

**Review by nirs — APPROVED** [review]

**Pull request merged**